File: mqttclients/dotnet/MQTTnet.Client.Extensions/ConnectionSettings.cs
using System.Diagnostics;
using System.Text;

namespace MQTTnet.Client.Extensions;

public enum AuthType
{
    X509,
    Basic
}

public class ConnectionSettings
{
    private const int Default_KeepAliveInSeconds = 60;
    private const string Default_CleanSession = "true";
    private const int Default_TcpPort = 8883;
    private const string Default_UseTls = "true";
    private const string Default_DisableCrl = "false";

    public string? HostName { get; set; }
    public string? ClientId { get; set; }
    public string? CertFile { get; set; }
    public string? KeyFile { get; set; }
    public string? KeyFilePassword { get; set; }

    public AuthType Auth
    {
        get => !string.IsNullOrEmpty(CertFile) ? AuthType.X509 : AuthType.Basic;
    }
    public string? UserName { get; set; }
    public string? Password { get; set; }
    public int KeepAliveInSeconds { get; set; }
    public bool CleanSession { get; set; }
    public int TcpPort { get; set; }
    public bool UseTls { get; set; }
    public string? CaFile { get; set; }
    public bool DisableCrl { get; set; }


    public ConnectionSettings()
    {
        TcpPort = Default_TcpPort;
        KeepAliveInSeconds = Default_KeepAliveInSeconds;
        UseTls = Default_UseTls == "true";
        DisableCrl = Default_DisableCrl == "true";
        CleanSession = Default_CleanSession == "true";
    }

    public static ConnectionSettings FromConnectionString(string cs) => new(cs);
    public ConnectionSettings(string cs) => ParseConnectionString(cs);

    public static ConnectionSettings CreateFromEnvVars(string? envFile = "")
    {
        if (string.IsNullOrEmpty(envFile))
        {
            envFile = ".env";
        }

        if (File.Exists(envFile))
        {
            Trace.TraceInformation("Loading environment variables from {envFile}" + new FileInfo(envFile).FullName);
            foreach (var line in File.ReadAllLines(envFile))
            {
                var parts = line.Split('=', StringSplitOptions.RemoveEmptyEntries);
                if (parts.Length != 2) continue;
                Environment.SetEnvironmentVariable(parts[0], parts[1]);
            }
        }
        else
        {
            Trace.TraceWarning($"EnvFile Not found in path {new DirectoryInfo(".").FullName} {envFile}");
        }
        static string ToUpperCaseFromPascalCase(string pascal) =>
            string.Concat(pascal.Select(x => Char.IsUpper(x) ? "_" + x : x.ToString())).ToUpper().TrimStart('_');

        static string Env(string name) =>
            Environment.GetEnvironmentVariable(ToUpperCaseFromPascalCase(name)) ?? string.Empty;

        string hostname = Env(nameof(HostName));

        ArgumentException.ThrowIfNullOrEmpty(hostname, nameof(hostname));

        return new ConnectionSettings
        {
            HostName = hostname,
            ClientId = Env(nameof(ClientId)),
            CertFile = Env(nameof(CertFile)),
            KeyFile = Env(nameof(KeyFile)),
            UserName = Env(nameof(UserName)),
            Password = Env(nameof(Password)),
            KeepAliveInSeconds = int.TryParse(Env(nameof(KeepAliveInSeconds)), out int keepAliveInSeconds) ? keepAliveInSeconds : Default_KeepAliveInSeconds,
            CleanSession = Env(nameof(CleanSession)) == "true",
            TcpPort = int.TryParse(Env(nameof(TcpPort)), out int tcpPort) ? tcpPort : Default_TcpPort,
            UseTls = string.IsNullOrEmpty(Env(nameof(UseTls))) || Env(nameof(UseTls)) == Default_UseTls,
            CaFile = Env(nameof(CaFile)),
            DisableCrl = Env(nameof(DisableCrl)) == "true"
        };
    }

    private static string GetStringValue(IDictionary<string, string> dict, string propertyName, string defaultValue = "")
    {
        string result = defaultValue;
        if (dict.TryGetValue(propertyName, out string? value))
        {
            result = value;
        }
        return result;
    }

    private static int GetPositiveIntValueOrDefault(IDictionary<string, string> dict, string propertyName, int defaultValue)
    {
        int result = defaultValue;
        if (dict.TryGetValue(propertyName, out string? stringValue))
        {
            if (int.TryParse(stringValue, out int intValue))
            {
                result = intValue;
            }
        }
        return result;
    }

    private void ParseConnectionString(string cs)
    {
        IDictionary<string, string> map = cs.ToDictionary(';', '=');
        HostName = GetStringValue(map, nameof(HostName));
        ClientId = GetStringValue(map, nameof(ClientId));
        KeyFile = GetStringValue(map, nameof(KeyFile));
        CertFile = GetStringValue(map, nameof(CertFile));
        UserName = GetStringValue(map, nameof(UserName));
        Password = GetStringValue(map, nameof(Password));
        KeepAliveInSeconds = GetPositiveIntValueOrDefault(map, nameof(KeepAliveInSeconds), Default_KeepAliveInSeconds);
        CleanSession = GetStringValue(map, nameof(CleanSession), Default_CleanSession) == "true";
        TcpPort = GetPositiveIntValueOrDefault(map, nameof(TcpPort), Default_TcpPort);
        UseTls = GetStringValue(map, nameof(UseTls), Default_UseTls) == "true";
        CaFile = GetStringValue(map, nameof(CaFile));
        DisableCrl = GetStringValue(map, nameof(DisableCrl), Default_DisableCrl) == "true";
        ArgumentNullException.ThrowIfNullOrEmpty(HostName);
    }

    private static void AppendIfNotEmpty(StringBuilder sb, string name, string val)
    {
        if (!string.IsNullOrEmpty(val))
        {
            sb.Append($"{name}={val};");
        }
    }

    public override string ToString()
    {
        var result = new StringBuilder();
        AppendIfNotEmpty(result, nameof(HostName), HostName!);
        AppendIfNotEmpty(result, nameof(TcpPort), TcpPort.ToString());
        AppendIfNotEmpty(result, nameof(UserName), UserName!);
        AppendIfNotEmpty(result, nameof(CleanSession), CleanSession.ToString());
        AppendIfNotEmpty(result, nameof(KeepAliveInSeconds), KeepAliveInSeconds.ToString());
        AppendIfNotEmpty(result, nameof(CertFile), CertFile!);
        AppendIfNotEmpty(result, nameof(KeyFile), KeyFile!);
        AppendIfNotEmpty(result, nameof(CaFile), CaFile!);
        AppendIfNotEmpty(result, nameof(ClientId), ClientId!);
        AppendIfNotEmpty(result, nameof(UseTls), UseTls.ToString());
        AppendIfNotEmpty(result, nameof(Auth), Auth!.ToString());
        result.Remove(result.Length - 1, 1);
        return result.ToString();
    }
}

File: mqttclients/ts/mqttjsClientExtensions/src/connectionSettings.ts
import { config } from 'dotenv';

enum AuthType {
    X509 = 'X509',
    Basic = 'Basic'
}

const defaultHostname = 'localhost';
const defaultKeepAliveInSeconds = 30;
const defaultCleanSession = true;
const defaultTcpPort = 8883;
const defaultUseTls = true;
const defaultDisableCrl = false;

export class ConnectionSettings {
    private _hostname = defaultHostname;
    private _clientId = '';
    private _certFile = '';
    private _keyFile = '';
    private _keyFilePassword = '';
    private _username = '';
    private _password = '';
    private _keepAliveInSeconds: number = defaultKeepAliveInSeconds;
    private _cleanSession: boolean = defaultCleanSession;
    private _tcpPort: number = defaultTcpPort;
    private _useTls: boolean = defaultUseTls;
    private _caFile = '';
    private _disableCrl: boolean = defaultDisableCrl;

    public static fromConnectionString(connectionString: string): ConnectionSettings {
        return ConnectionSettings.parseConnectionString(connectionString);
    }

    public static createFromEnvVars(envFilePath = '.env'): ConnectionSettings {
        // Load environment variables from .env file using the dotenv package
        const envConfig = config({ path: envFilePath });

        if (!envConfig.parsed?.MQTT_HOST_NAME) {
            throw new Error('MQTT_HOST_NAME environment variable is not set');
        }

        if (envConfig.parsed?.MQTT_PASSWORD && !envConfig.parsed?.MQTT_USERNAME) {
            throw new Error('MQTT_USERNAME environment variable is required if MQTT_PASSWORD is set');
        }

        const cs = new ConnectionSettings();

        cs._hostname = envConfig.parsed?.MQTT_HOST_NAME || defaultHostname;
        cs._tcpPort = Number(envConfig.parsed?.MQTT_TCP_PORT) || defaultTcpPort;
        cs._useTls = Boolean(envConfig.parsed?.MQTT_USE_TLS === undefined ? defaultUseTls : envConfig.parsed?.MQTT_USE_TLS);
        cs._cleanSession = Boolean(envConfig.parsed?.MQTT_CLEAN_SESSION === undefined ? defaultCleanSession : envConfig.parsed?.MQTT_CLEAN_SESSION);
        cs._keepAliveInSeconds = Number(envConfig.parsed?.MQTT_KEEP_ALIVE_IN_SECONDS) || defaultKeepAliveInSeconds;
        cs._clientId = envConfig.parsed?.MQTT_CLIENT_ID || '';
        cs._username = envConfig.parsed?.MQTT_USERNAME || '';
        cs._password = envConfig.parsed?.MQTT_PASSWORD || '';
        cs._certFile = envConfig.parsed?.MQTT_CERT_FILE || '';
        cs._keyFile = envConfig.parsed?.MQTT_KEY_FILE || '';
        cs._caFile = envConfig.parsed?.MQTT_CA_FILE || '';

        return cs;
    }

    public static parseConnectionString(_connectionString: string): ConnectionSettings {
        const cs = new ConnectionSettings();

        // cs.tcpPort = Number(envConfig.parsed?.MQTT_TCP_PORT) || 8883;
        // cs.useTls = Boolean(envConfig.parsed?.MQTT_USE_TLS === undefined ? true : envConfig.parsed?.MQTT_USE_TLS);
        // cs.cleanSession = Boolean(envConfig.parsed?.MQTT_CLEAN_SESSION === undefined ? true : envConfig.parsed?.MQTT_CLEAN_SESSION);
        // cs.keepAliveInSeconds = Number(envConfig.parsed?.MQTT_KEEP_ALIVE_IN_SECONDS) || 30;
        // cs.clientId = envConfig.parsed?.MQTT_CLIENT_ID || '';
        // cs.username = envConfig.parsed?.MQTT_USERNAME || '';
        // cs.password = envConfig.parsed?.MQTT_PASSWORD || '';
        // cs.certFile = envConfig.parsed?.MQTT_CERT_FILE || '';
        // cs.keyFile = envConfig.parsed?.MQTT_KEY_FILE || '';
        // cs.caFile = envConfig.parsed?.MQTT_CA_FILE || '';

        return cs;
    }

    public get hostname(): string {
        return this._hostname;
    }

    public get clientId(): string {
        return this._clientId;
    }

    public set clientId(value: string) {
        this._clientId = value;
    }

    public get certFile(): string {
        return this._certFile;
    }

    public set certFile(value: string) {
        this._certFile = value;
    }

    public get keyFile(): string {
        return this._keyFile;
    }

    public set keyFile(value: string) {
        this._keyFile = value;
    }

    public get keyFilePassword(): string {
        return this._keyFilePassword;
    }

    public set keyFilePassword(value: string) {
        this._keyFilePassword = value;
    }

    public get auth(): AuthType {
        return this._certFile ? AuthType.X509 : AuthType.Basic;
    }

    public get username(): string {
        return this._username;
    }

    public set username(value: string) {
        this._username = value;
    }

    public get password(): string {
        return this._password;
    }

    public set password(value: string) {
        this._password = value;
    }

    public get keepAliveInSeconds(): number {
        return this._keepAliveInSeconds;
    }

    public set keepAliveInSeconds(value: number) {
        this._keepAliveInSeconds = value;
    }

    public get cleanSession(): boolean {
        return this._cleanSession;
    }

    public set cleanSession(value: boolean) {
        this._cleanSession = value;
    }

    public get tcpPort(): number {
        return this._tcpPort;
    }

    public set tcpPort(value: number) {
        this._tcpPort = value;
    }

    public get useTls(): boolean {
        return this._useTls;
    }

    public set useTls(value: boolean) {
        this._useTls = value;
    }

    public get caFile(): string {
        return this._caFile;
    }

    public set caFile(value: string) {
        this._caFile = value;
    }

    public get disableCrl(): boolean {
        return this._disableCrl;
    }

    public set disableCrl(value: boolean) {
        this._disableCrl = value;
    }
}
